action/protocol/staking/protocol.go

```diff
@@ -684,3 +684,37 @@ func writeCandCenterStateToStateDB(sm protocol.StateManager, name, op, owners Ca
 	_, err := sm.PutState(owners, protocol.NamespaceOption(CandsMapNS), protocol.KeyOption(_ownerKey))
 	return err
 }
+
+// isSelfStakeBucket returns true if the bucket is self-stake bucket and not expired
+func isSelfStakeBucket(featureCtx protocol.FeatureCtx, csm CandidateStateManager, index uint64) (bool, error) {
+	// bucket index should be settled in one of candidates
+	selfStake := csm.ContainsSelfStakingBucket(index)
```

Comment:
i recall we discussed, if an endorsement is withdraw, we'll change the `selfStakeBucketId` of currently endorsed candidate to `max.Uint64`?
in that case, `ContainsSelfStakingBucket` is still valid and reflect the accurate status?

Reply:
only update `exipreHeight` when endorsement withdrawal, the clean of `selfStakeBucketId` should be at next `ChangeDelegate` action of the endorse bucket